File: AzimMuradov/Term4/Block3/Task5/core/src/main/kotlin/deanery/set/LazySet.kt
package deanery.set

import java.util.concurrent.atomic.AtomicInteger
import kotlin.concurrent.withLock
import deanery.set.LazySetNode as Node


class LazySet<T : Any> : ConcurrentSet<T> {

    private val head = Node<T>(
        key = Int.MIN_VALUE,
        next = Node(key = Int.MAX_VALUE)
    )


    private val _count = AtomicInteger()

    override val count: Int get() = _count.get()


    override fun contains(element: T): Boolean {
        val key: Int = element.hashCode()
        val curr = generateSequence(head) { if (it.key < key) it.next else null }.last()
        return curr.key == key && !curr.mark
    }

    override fun add(element: T) {
        val key: Int = element.hashCode()

        while (true) {
            findWindow(key).withLock { (prev, curr) ->
                if (validate(prev, curr)) {
                    if (curr.key != key) {
                        prev.next = Node(element, next = curr)
                        _count.incrementAndGet()
                    }
                    return
                }
            }
        }
    }

    override fun remove(element: T) {
        val key: Int = element.hashCode()

        while (true) {
            findWindow(key).withLock { (prev, curr) ->
                if (validate(prev, curr)) {
                    if (curr.key == key) {
                        curr.mark = true
                        prev.next = curr.next
                        _count.decrementAndGet()
                    }
                    return
                }
            }
        }
    }


    private fun findWindow(key: Int) = generateSequence(seed = head to head.next) { (_, curr) ->
        if (curr.key < key) {
            curr to curr.next
        } else {
            null
        }
    }.last()


    companion object {

        private fun validate(prev: Node<*>, curr: Node<*>): Boolean =
            !prev.mark && !curr.mark && prev.next == curr

        private inline fun <T : Any> Pair<Node<T>, Node<T>>.withLock(b: (Pair<Node<T>, Node<T>>) -> Unit) =
            first.withLock { second.withLock { b(this) } }
    }
}

File: VyacheslavBuchin/Term4/Block3/CreditSystem/ConcurrentSet/LazySet.cs
namespace ConcurrentSet;

public class LazySet<T> : ISet<T>
{
    private readonly Node<T> _tail = new(int.MaxValue);
    private readonly Node<T> _head = new(int.MinValue);
    private readonly ReaderWriterLockSlim _countLock = new();
    private int _count = 0;

    public LazySet()
    {
        _head.Next = _tail;
    }

    private bool Validate(Node<T> pred, Node<T> curr)
    {
        return !pred.Marked && !curr.Marked && pred.Next == curr;
    }

    public int Count()
    {
        _countLock.EnterReadLock();
        var count = _count;
        _countLock.ExitReadLock();
        return count;
    }

    public bool Add(T item)
    {
        var key = item!.GetHashCode();
        while (true)
        {
            var pred = _head;
            var curr = _head.Next;
            while (curr.Key < key)
            {
                pred = curr;
                curr = curr.Next;
            }

            pred.Lock();
            try
            {
                curr.Lock();
                try
                {
                    if (Validate(pred, curr))
                    {
                        if (curr.Key == key)
                        {
                            return false;
                        }
                        else
                        {
                            var node = new Node<T>(item) { Next = curr };
                            pred.Next = node;
                            _countLock.EnterWriteLock();
                            ++_count;
                            _countLock.ExitWriteLock();
                            return true;
                        }
                    }
                }
                finally
                {
                    curr.Unlock();
                }
            }
            finally
            {
                pred.Unlock();
            }
        }
    }

    public bool Remove(T item)
    {
        var key = item!.GetHashCode();
        while (true)
        {
            var pred = _head;
            var curr = _head.Next;
            while (curr.Key < key)
            {
                pred = curr;
                curr = curr.Next;
            }

            pred.Lock();
            try
            {
                curr.Lock();
                try
                {
                    if (Validate(pred, curr))
                    {
                        if (curr.Key != key)
                        {
                            return false;
                        }
                        else
                        {
                            curr.Marked = true;
                            pred.Next = curr.Next;
                            _countLock.EnterWriteLock();
                            --_count;
                            _countLock.ExitWriteLock();
                            return true;
                        }
                    }
                }
                finally
                {
                    curr.Unlock();
                }
            }
            finally
            {
                pred.Unlock();
            }
        }
    }

    public bool Contains(T item)
    {
        var key = item!.GetHashCode();
        var curr = _head;
        while (curr.Key < key)
            curr = curr.Next;
        return curr.Key == key && !curr.Marked;
    }
}
